### PR TITLE
make pagerank faster

### PR DESCRIPTION
**Pull request by afternone**
Here are the benchmarks:
```julia
julia> g = DiGraph(10000, 10000*25) # sparse graph
{10000, 250000} directed simple Int64 graph

julia> @benchmark pagerank_opt(g)
BenchmarkTools.Trial:
  memory estimate:  234.83 KiB
  allocs estimate:  11
  --------------
  minimum time:     24.690 ms (0.00% GC)
  median time:      25.774 ms (0.00% GC)
  mean time:        26.514 ms (0.06% GC)
  maximum time:     40.263 ms (0.00% GC)
  --------------
  samples:          189
  evals/sample:     1

julia> @benchmark pagerank(g)
BenchmarkTools.Trial:
  memory estimate:  16.49 MiB
  allocs estimate:  93
  --------------
  minimum time:     32.841 ms (0.00% GC)
  median time:      37.650 ms (7.79% GC)
  mean time:        37.562 ms (7.41% GC)
  maximum time:     46.090 ms (12.23% GC)
  --------------
  samples:          134
  evals/sample:     1

julia> g = DiGraph(10000, 10000*2500) # dense graph
{10000, 25000000} directed simple Int64 graph

julia> @benchmark pagerank_opt(g)
BenchmarkTools.Trial:
  memory estimate:  234.83 KiB
  allocs estimate:  11
  --------------
  minimum time:     1.235 s (0.00% GC)
  median time:      1.251 s (0.00% GC)
  mean time:        1.250 s (0.00% GC)
  maximum time:     1.261 s (0.00% GC)
  --------------
  samples:          5
  evals/sample:     1

julia> @benchmark pagerank(g)
BenchmarkTools.Trial:
  memory estimate:  1.49 GiB
  allocs estimate:  87
  --------------
  minimum time:     4.557 s (10.71% GC)
  median time:      4.704 s (12.65% GC)
  mean time:        4.704 s (12.65% GC)
  maximum time:     4.851 s (14.47% GC)
  --------------
  samples:          2
  evals/sample:     1

julia> g = barabasi_albert(100000, 100, is_directed=true)
{100000, 9990000} directed simple Int64 graph

julia> @benchmark pagerank_opt(g)
BenchmarkTools.Trial:
  memory estimate:  2.31 MiB
  allocs estimate:  519
  --------------
  minimum time:     1.572 s (0.00% GC)
  median time:      1.574 s (0.00% GC)
  mean time:        1.578 s (0.00% GC)
  maximum time:     1.592 s (0.00% GC)
  --------------
  samples:          4
  evals/sample:     1

julia> @benchmark pagerank(g)
BenchmarkTools.Trial:
  memory estimate:  625.04 MiB
  allocs estimate:  105
  --------------
  minimum time:     2.951 s (20.73% GC)
  median time:      3.120 s (21.82% GC)
  mean time:        3.120 s (21.82% GC)
  maximum time:     3.288 s (22.81% GC)
  --------------
  samples:          2
  evals/sample:     1
```

### CODE REVIEW

**Comment by sbromberger** [comment]
Hi @afternone - thanks for this. I've been out of town for a few days but am getting back. I'll take a look in the next couple days, but paging @jpfairbanks as well.

**Comment by sbromberger** [comment]
Can we get this ready for merge?

File: src/centrality/pagerank.jl [hunk 1]
```diff
@@ -12,34 +12,40 @@ is not reached within `n` iterations.
...
 
-@traitfn function pagerank(g::::IsDirected, α=0.85, n=100::Integer, ϵ=1.0e-6)
-    A = adjacency_matrix(g, Float64; dir=:in)
-    S = vec(sum(A, 1))
-    S = 1 ./ S
-    S[find(S .== Inf)] = 0.0
-    M = A'  # need a separate line due to bug #17456 in julia
-    # scaling the adjmat to stochastic adjacency matrix
-    M = (Diagonal(S) * M)'
+@traitfn function pagerank(g::AbstractGraph, α=0.85, n=100::Integer, ϵ=1.0e-6)
```
Comment (sbromberger) [inline src/centrality/pagerank.jl]: Is this an enhancement to pagerank to accept undirected graphs, or should we still restrict to DiGraphs?

Comment (jpfairbanks) [reply src/centrality/pagerank.jl]: The algorithm works for undirected graphs, so I don't see a good reason for restricting it.

**Comment by sbromberger** [comment]
I'm going to merge this given #792's promise to get code coverage back up. @jpfairbanks - if we're not at 100% after both PRs are merged, would you please help w/ codecov?

**Pull request merged**